### ISSUE TITLE
TypeScript doesn't resolve definitions for submodules

### ISSUE THREAD

**Issue by MaximDevoir** [octokit/graphql.js#52]
Opened a related issue in the Pika builder repo: [issue 83](https://github.com/pikapkg/builders/issues/83). This is something that should be fixed in the `plugin-ts-standard-pkg` repository. In the meantime, I think we need a patch solution here.

**The Issue**
In graphql 4.0.0, submodule definition files were generated to 'dist-types' by the `pika` builder. Developers will only get access to the the definitions for `index.d.ts` because of where and how the source files were generated.

**Solutions**

- Add all the submodules as exports to `index.ts`. This requires a major version bump when the exports are eventually removed from `index.ts`. This pattern could also run into naming collisions. 
- Copy the generated `d.ts` files to `node-src` directory. This should probably be done in a post-build script.

**Comment by MaximDevoir** [comment]
I'm working on migrating some old `graphql` definitions in Probot to this repository. So whatever solution you would like to move forward with, I'll add that to the PR.

**Comment by gr2m** [comment]
I'm sorry I don't quite understand the problem? The package.json file has the `"types"` key which points to `dist-types/index.d.ts`

https://unpkg.com/browse/@octokit/graphql@4.0.0/package.json

So the types should be exported correctly when importing `{ graphql }` from `@octokit/graphql`?

I think I'm missing your point, could you maybe provide an example?

**Comment by MaximDevoir** [comment]
Yeah, they are needed in some areas in the probot/probot repository.

Such as importing [with-defaults](https://github.com/probot/probot/blob/f457477e505bc9f0463ae1cf023f13067f0fca11/src/github/graphql.ts#L18)

**Comment by gr2m** [comment]
Ah I see what you mean now. And I remember, that is actually the reason why Probot was not yet updated to the latest `@octokit/graphql.js` version: Probot is using internal APIs which are no longer exposed.

If I recall correctly, we use the internal "WithDefaults" method because we needed to be able to pass in Probots `octokit.request` instance which has plugins applied, e.g. for throttling and retries.

I guess we could export the `withDefaults` method but add a comment making clear that it is considered an internal API and that it can break at any time, so consumers should pin the version. Should be as simple as replacing


What do you think?

**Comment by gr2m** [comment]
Maybe we could add a new export which only allows to set a custom request method, but not the default options, somthing like this

```js
import { withDefaults } from "./with-defaults";

export function withCustomRequest (request) {
  return withDefaults(request, {
    headers: {
      "user-agent": `octokit-graphql.js/${VERSION} ${getUserAgent()}`
    },
    method: "POST",
    url: "/graphql"
  });
}
```

We could make that an official, public API and add documentation for it in the README.md

**Comment by gr2m** [comment]
I'm going to build this now ☝️ I'm not 100% convinced that it is necessary but it will unblock Probot to use the latest version without waiting for the next major release when probot will be based on [@octokit/core](https://github.com/octokit/core.js)

**Comment by gr2m** [comment]
The new `withCustomRequest` method has been shipped in [v4.2.0](https://github.com/octokit/graphql.js/releases/tag/v4.2.0)

This should resolve https://github.com/probot/probot/pull/1000 and unblock https://github.com/probot/probot/pull/986. Would you like to send an according PR to Probot?

**Comment by gr2m** [comment]
resolved via https://github.com/probot/probot/pull/1006
